src/main/java/games/strategy/engine/framework/startup/login/V2Authenticator.java

```diff
@@ -0,0 +1,223 @@
+    // https://stackoverflow.com/questions/27622625/securerandom-with-nativeprng-vs-sha1prng
+    final SecureRandom secureRandom = new SecureRandom();
+
+    final byte[] bytes = new byte[length];
+    secureRandom.nextBytes(bytes);
+    return bytes;
+  }
+
+  /**
+   * Indicates the specified response can be processed by this authenticator.
+   *
+   * @param response The response received from the client node; must not be {@code null}.
+   *
+   * @return {@code true} if the specified response can be processed by this authenticator; otherwise {@code false}.
+   */
+  static boolean canProcessResponse(final Map<String, String> response) {
+    return response.containsKey(ResponsePropertyNames.DIGEST);
+  }
+
+  @VisibleForTesting
+  interface ResponsePropertyNames {
+    String DIGEST = "authenticator/v2/digest";
+  }
+
+  /**
+   * Creates a response to the specified challenge for the client node to send to the server node.
+   *
+   * @param password The game password; must not be {@code null}.
+   * @param challenge The challenge as a collection of properties; must not be {@code null}.
+   *
+   * @return The response as a collection of properties to be added to the message the client node sends back to the
+   *         server node; never {@code null}.
+   *
+   * @throws AuthenticationException If an error occurs while parsing the challenge or creating the response.
+   */
+  static Map<String, String> newResponse(
+      final String password,
+      final Map<String, String> challenge)
+      throws AuthenticationException {
+    final byte[] nonce = decodeOptionalProperty(challenge, ChallengePropertyNames.NONCE);
+    final byte[] salt = decodeOptionalProperty(challenge, ChallengePropertyNames.SALT);
+    if (nonce == null || salt == null) {
+      return Collections.emptyMap();
+    }
+
+    try {
+      return Maps.newHashMap(ImmutableMap.<String, String>builder()
+          .put(encodeProperty(ResponsePropertyNames.DIGEST, digest(password, salt, nonce)))
+          .build());
+    } catch (final GeneralSecurityException e) {
+      throw new AuthenticationException("security framework failure when creating response", e);
+    }
+  }
+
+  @VisibleForTesting
+  static byte[] decodeOptionalProperty(
+      final Map<String, String> properties,
+      final String name)
+      throws AuthenticationException {
+    final String value = properties.get(name);
+    if (value == null) {
+      return null;
+    }
+
+    try {
+      return Base64.getDecoder().decode(value);
+    } catch (final IllegalArgumentException e) {
+      throw new AuthenticationException(String.format("malformed value for property '%s'", name), e);
+    }
+  }
+
+  private static byte[] digest(
+      final String password,
+      final byte[] salt,
+      final byte[] nonce)
+      throws GeneralSecurityException {
+    final Mac mac = Mac.getInstance("HmacSHA512");
+    mac.init(newSecretKey(password, salt));
+    mac.update(nonce);
+    return mac.doFinal();
```

Comment:
This line can be merged with the preceding line...
`mac.doFinal(byte[])` is essentially calling `mac.update(bytre[]); mac.dofinal()`